vmgr/vmgr_listtape.man
.\" Copyright (C) 2000-2002 by CERN/IT/PDP/DM
.\" All rights reserved
.\"
.TH VMGR_LISTTAPE 3 "$Date: 2002/02/07 06:46:24 $" CASTOR "vmgr Library Functions"
.SH NAME
vmgr_listtape \- list all existing tape entries in the CASTOR Volume Manager
.SH SYNOPSIS
.B #include <sys/types.h>
.br
\fB#include "vmgr_api.h"\fR
.sp
.BI "struct vmgr_tape_info *vmgr_listtape (char *" vid ,
.BI "char *" pool_name ,
.BI "int " flags ,
.BI "vmgr_list *" listp )
.SH DESCRIPTION
.B vmgr_listtape
lists all existing tape entries in the CASTOR Volume Manager.
.P
Please note that users of the vmgr client API are encouraged to use the
.B vmgr_listtape_byte_u64
function instead of
.B vmgr_listtape.
The former supports tape sizes up to 16 Exibytes whereas the latter only
supports tape sizes up to 2 Teribytes.
If the
.B vmgr_listtape
function attempts to list a tape with more than 2 Teribytes of free space,
then the function will store a value of exactly 2 Teribytes into the
.I estimated_free_space
member of the structure pointed to by the function's return value.
The
.B vmgr_listtape_byte_u64
function defines free space to be in bytes and represented by a 64-bit unsigned
integer.
The
.B vmgr_listtape
defines free space to be in kibibytes and represented by a
signed 32-bit integer.
The
.B vmgr_listtape_byte_u64
function has been introduced to help represent all CASTOR sizes in the same
way.
.TP
.I vid
if not NULL, restricts the list to the media having this volume visual identifier.
.TP
.I pool_name
if not NULL, restricts the list to the tapes belonging to this pool.
.TP
.I flags
may be one of the following constants:
.RS
.TP
.B VMGR_LIST_BEGIN
the first call for a volume must have this flag set to allocate buffers and
initialize pointers.
.TP
.B VMGR_LIST_CONTINUE
all the following calls must have this flag set.
.TP
.B VMGR_LIST_END
final call to terminate the list and free resources.
.RE
.SH RETURN VALUE
This routine returns a pointer to a structure containing the current tape entry
if the operation was successful or NULL if all entries have been returned
or if the operation failed. In the latter case,
.B serrno
is set appropriately.
.SH ERRORS
.TP 1.3i
.B ENOMEM
memory could not be allocated for the output buffer.
.TP
.B EFAULT
.I listp
is a NULL pointer.
.TP
.B EINVAL
The length of
.I pool_name
exceeds
.BR CA_MAXPOOLNAMELEN .
.TP
.B SENOSHOST
Host unknown.
.TP
.B SENOSSERV
Service unknown.
.TP
.B SECOMERR
Communication error.
.TP
.B EVMGRNACT
Volume manager is not running or is being shutdown.
.SH SEE ALSO
.BR Castor_limits(4) ,
.BR vmgr_entertape(3)
.SH AUTHOR
\fBCASTOR\fP Team <castor.support@cern.ch>


vmgr/vmgr_listtape.c
/*
 * Copyright (C) 2000-2002 by CERN/IT/PDP/DM
 * All rights reserved
 */
 
/*      vmgr_listtape - list all existing tapes */

#include <errno.h>
#include <stdlib.h>
#include <sys/types.h>
#include <unistd.h>
#include <netinet/in.h>
#include "marshall.h"
#include "serrno.h"
#include "vmgr_api.h"
#include "vmgr.h"

struct vmgr_tape_info *
vmgr_listtape(char *vid, char *pool_name, int flags, vmgr_list *listp)
{
	int bol = 0;
	int c;
	char func[14];
	gid_t gid;
	int listentsz = sizeof(struct vmgr_tape_info);
	struct vmgr_tape_info *lp;
	int msglen;
	int nbentries;
	char *q;
	char *rbp;
	char repbuf[LISTBUFSZ+4];
	char *sbp;
	char sendbuf[REQBUFSZ];
	struct vmgr_api_thread_info *thip;
	uid_t uid;

        strncpy (func, "vmgr_listtape", 14);
        if (vmgr_apiinit (&thip))
                return (NULL);
        uid = geteuid();
        gid = getegid();

	if (! listp) {
		serrno = EFAULT;
		return (NULL);
	}

	if (flags == VMGR_LIST_BEGIN) {
		memset (listp, 0, sizeof(vmgr_list));
		listp->fd = -1;
		if ((listp->buf = malloc (LISTBUFSZ)) == NULL) {
			serrno = ENOMEM;
			return (NULL);
		}
		bol = 1;
	}
	if (listp->nbentries == 0 && listp->eol	/* all entries have been listed */
	    && flags != VMGR_LIST_END)
		return (NULL);

	if (listp->nbentries == 0	/* no data in the cache */
	    || flags == VMGR_LIST_END) {

		/* Build request header */

		sbp = sendbuf;
		marshall_LONG (sbp, VMGR_MAGIC2);
		if (flags == VMGR_LIST_END) {
			marshall_LONG (sbp, VMGR_ENDLIST);
		} else {
			marshall_LONG (sbp, VMGR_LISTTAPE);
		}
		q = sbp;        /* save pointer. The next field will be updated */
		msglen = 3 * LONGSIZE;
		marshall_LONG (sbp, msglen);

		/* Build request body */
	 
		marshall_LONG (sbp, uid);
		marshall_LONG (sbp, gid);
		marshall_WORD (sbp, listentsz);
		if (vid) {
			marshall_STRING (sbp, vid);
		} else {
			marshall_STRING (sbp, "");
		}
		if (pool_name) {
			marshall_STRING (sbp, pool_name);
		} else {
			marshall_STRING (sbp, "");
		}
		marshall_WORD (sbp, bol);
	 
		msglen = sbp - sendbuf;
		marshall_LONG (q, msglen);	/* update length field */

		while ((c = send2vmgr (&listp->fd, sendbuf, msglen,
		    repbuf, sizeof(repbuf))) && serrno == EVMGRNACT)
			sleep (RETRYI);
		if (c < 0)
			return (NULL);
		if (flags == VMGR_LIST_END) {
			if (listp->buf)
				free (listp->buf);
			return (NULL);
		}
		rbp = repbuf;
		unmarshall_WORD (rbp, nbentries);
		if (nbentries == 0)
			return (NULL);		/* end of list */

		/* unmarshall reply into vmgr_tape_info structures */

		listp->nbentries = nbentries;
		lp = (struct vmgr_tape_info *) listp->buf;
		while (nbentries--) {
			unmarshall_STRING (rbp, lp->vid);
			unmarshall_STRING (rbp, lp->vsn);
			unmarshall_STRING (rbp, lp->library);
			unmarshall_STRING (rbp, lp->density);
			unmarshall_STRING (rbp, lp->lbltype);
			unmarshall_STRING (rbp, lp->model);
			unmarshall_STRING (rbp, lp->media_letter);
			unmarshall_STRING (rbp, lp->manufacturer);
			unmarshall_STRING (rbp, lp->sn);
			unmarshall_WORD (rbp, lp->nbsides);
			unmarshall_TIME_T (rbp, lp->etime);
			unmarshall_WORD (rbp, lp->side);
			unmarshall_STRING (rbp, lp->poolname);
			unmarshall_LONG (rbp, lp->estimated_free_space);
			unmarshall_LONG (rbp, lp->nbfiles);
			unmarshall_LONG (rbp, lp->rcount);
			unmarshall_LONG (rbp, lp->wcount);
			unmarshall_STRING (rbp, lp->rhost);
			unmarshall_STRING (rbp, lp->whost);
			unmarshall_LONG (rbp, lp->rjid);
			unmarshall_LONG (rbp, lp->wjid);
			unmarshall_TIME_T (rbp, lp->rtime);
			unmarshall_TIME_T (rbp, lp->wtime);
			unmarshall_LONG (rbp, lp->status);
			lp++;
		}
		unmarshall_WORD (rbp, listp->eol);
	}
	lp = ((struct vmgr_tape_info *) listp->buf) + listp->index;
	listp->index++;
	if (listp->index >= listp->nbentries) {	/* must refill next time */
		listp->index = 0;
		listp->nbentries = 0;
	}
	return (lp);
}
